src/MOI_wrapper.jl

```diff
@@ -0,0 +1,2708 @@
+)
+    if haskey(model.sos_constraint_info, key.value)
+        return model.sos_constraint_info[key.value]
+    end
+    throw(MOI.InvalidIndex(key))
+end
+
+_sos_type(::MOI.SOS1) = :SOS1
+_sos_type(::MOI.SOS2) = :SOS2
+
+function MOI.is_valid(
+    model::Optimizer,
+    c::MOI.ConstraintIndex{MOI.VectorOfVariables, S}
+) where {S}
+    info = get(model.sos_constraint_info, c.value, nothing)
+    if info === nothing
+        return false
+    else
+        return typeof(info.set) == S
+    end
+end
+
+function MOI.add_constraint(
+    model::Optimizer, f::MOI.VectorOfVariables, s::SOS
+)
+    columns = Int[_info(model, v).column for v in f.variables]
+    add_sos!(model.inner, _sos_type(s), columns, s.weights)
+    model.last_constraint_index += 1
+    index = MOI.ConstraintIndex{MOI.VectorOfVariables, typeof(s)}(model.last_constraint_index)
+    model.sos_constraint_info[index.value] = ConstraintInfo(
+        length(model.sos_constraint_info) + 1, s
+    )
+    return index
+end
+
+function MOI.delete(
+    model::Optimizer, c::MOI.ConstraintIndex{MOI.VectorOfVariables, <:SOS}
+)
+    row = _info(model, c).row
+    del_sos!(model.inner, [Cint(row)])
+    for (key, info) in model.sos_constraint_info
+        if info.row > row
+            info.row -= 1
+        end
+    end
+    delete!(model.sos_constraint_info, c.value)
+    return
+end
+
+function MOI.get(
+    model::Optimizer, ::MOI.ConstraintName,
+    c::MOI.ConstraintIndex{MOI.VectorOfVariables, <:Any}
+)
+    return _info(model, c).name
+end
+
+function MOI.set(
+    model::Optimizer, ::MOI.ConstraintName,
+    c::MOI.ConstraintIndex{MOI.VectorOfVariables, <:Any}, name::String
+)
+    info = _info(model, c)
+    if !isempty(info.name) && model.name_to_constraint_index !== nothing
+        delete!(model.name_to_constraint_index, info.name)
+    end
+    info.name = name
+    if model.name_to_constraint_index === nothing || isempty(name)
+        return
+    end
+    if haskey(model.name_to_constraint_index, name)
+        model.name_to_constraint_index = nothing
+    else
+        model.name_to_constraint_index[name] = c
+    end
+    return
+end
+
+function MOI.get(
+    model::Optimizer, ::MOI.ConstraintSet,
+    c::MOI.ConstraintIndex{MOI.VectorOfVariables, S}
+) where {S <: SOS}
+    full_matrix, _ = get_sos_matrix(model.inner)
+    line = full_matrix[_info(model, c).row,:] #sparse vec
+    return S(line.nzval)
+end
+
+
+function MOI.get(
+    model::Optimizer, ::MOI.ConstraintFunction,
+    c::MOI.ConstraintIndex{MOI.VectorOfVariables, S}
+) where {S <: SOS}
+    full_matrix, _ = get_sos_matrix(model.inner)
+    line = full_matrix[_info(model, c).row,:] #sparse vec
+    cols = line.nzind
+    return MOI.VectorOfVariables(
+        [model.variable_info[CleverDicts.LinearIndex(i)].index for i in cols]
+    )
+end
+
+###
+### Optimize methods.
+###
+
+function MOI.optimize!(model::Optimizer)
+    # Note: although Gurobi will call update regardless, we do it now so that

```

Comment:
Gurobi -> Xpress?

Reply:
Haven't got this far currently still at line 1718